programs/ssl/ssl_base64_dump.c

```diff
@@ -0,0 +1,953 @@
+
+#define TRANSFORM_RANDBYTE_LEN  64
+
+/*
+ * A macro that prevents from out of the ssl buffer range.
+ */
+const char buf_ln_err[] = "Buffer has not enough data to complete the parsing\n";
+#define CHECK_SSL_END( LEN )            \
+do                                      \
+{                                       \
+    if( end - ssl <  (int)( LEN ) )     \
+    {                                   \
+        printf_err( "%s", buf_ln_err ); \
+        return;                         \
+    }                                   \
+} while ( 0 )
+
+
+/*
+ * Global values
+ */
+FILE *b64_file = NULL;                  /* file with base64 codes to deserialize */
+char conf_keep_peer_certificate = 1;    /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE from mbedTLS configuration */
+char conf_dtls_proto = 1;               /* MBEDTLS_SSL_PROTO_DTLS from mbedTLS configuration */
+char debug = 0;                         /* flag for debug messages */
+
+/*
+ * Base printing functions
+ */
+void print_version( )
+{
+    printf( "%s v%d.%d\n", PROG_NAME, VER_MAJOR, VER_MINOR );
+}
+
+void print_usage( )
+{
+    print_version();
+    printf( "\nThis program is used to deserialize an mbedTLS SSL session from the base64 code provided\n"
+            "in the text file. The program can deserialize many codes from one file, but they must be\n"
+            "separated, e.g. by a newline.\n\n"
+            "Note: This program has been designed to work with mbedTLS v2.21.0 and may have some problems\n"
+            "      with other versions.\n\n" );
+    printf(
+        "Usage:\n"
+        "\t-f path            - Path to the file with base64 code\n"
+        "\t-v                 - Show version\n"
+        "\t-h                 - Show this usage\n"
+        "\t-d                 - Print more information\n"
+        "\t--keep-peer-cert=0 - Use this option if you know that the mbedTLS library\n"
+        "\t                     has been compiled with the MBEDTLS_SSL_KEEP_PEER_CERTIFICATE\n"
+        "\t                     flag. You can also use it if there are some problems with reading\n"
+        "\t                     the information about certificate\n"
+        "\t--dtls-protocol=0  - Use this option if you know that the mbedTLS library\n"
+        "\t                     has been compiled without the MBEDTLS_SSL_PROTO_DTLS flag\n"
+        "\n"
+    );
+}
+
+void printf_dbg( const char *str, ... )
+{
+    if( debug )
+    {
+        va_list args;
+        va_start( args, str );
+        printf( "debug: ");
+        vprintf( str, args );
+        fflush( stdout );
+        va_end( args );
+    }
+}
+
+void printf_err( const char *str, ... )
+{
+    va_list args;
+    va_start( args, str );
+    fflush( stdout );
+    fprintf( stderr, "ERROR: ");
+    vfprintf( stderr, str, args );
+    fflush( stderr );
+    va_end( args );
+}
+
+/*
+ * Exit from the program in case of error
+ */
+void error_exit()
+{
+    if( b64_file )
```

Comment:
If I'm not mistaken, there's a custom to be specific with NULL checks, so I would suggest 
```suggestion
    if( b64_file != NULL )
```
or, if you're fancy,
```suggestion
    if( NULL != b64_file )
```

Reply:
Actually, it's not only fancy, but it's also used as a good practice to reduce the probability of error :). For example, when you make a typo and don't write the double '=='. Then 'if (val = ENUM)' will cause runtime error and unforeseen behavior, while if (ENUM = val) will cause only compiling error

Reply:
I do know that (that's why I said it's fancy), but considering the amount of reviews and the approach to code in the library, such error is highly unlikely, and the approach is not generally used (not sure if it's used anywhere in the library).